src/merkle_tree.rs

```diff
@@ -128,15 +138,29 @@ impl MerkleTree {
     }
 
     /// returns the merkle proof for the element given its index in the leaves array. If the element is not present in the tree it will return an empty proof
-    pub fn proof(&self, elem_index: usize) -> Vec<String> {
-        let mut current = elem_index + self.nodes.len() / 2;
-        let mut proof = Vec::new();
-        assert!(self.is_leaf(current));
-        while current != self.root_index.unwrap() {
-            proof.push(self.nodes[MerkleTree::sibling_index(current)].to_string());
-            current = MerkleTree::parent_index(current);
+    pub fn proof<T>(&self, element: T) -> Vec<String>
+    where
+        T: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        element.hash(&mut hasher);
+        let hash = hasher.finish().to_string();
+        match self
```

Comment:
hace lo mismo que leaf_index_of? el chequeo lo hace con todos los nodos, pero no tendría que ser sobre las hojas nomás?

Reply:
Lo estuve viendo y mas abajo intercambiamos ese indice por el del padre, asi que necesitamos que sea el indice de la totalidad de nodos por que si no, no va a tomar al padre de forma correcta, igualmente lo podemos ver pero no lo cambie

Reply:
oh, you're right!